woocommerce/woocommerce
# Issue/31562 filter coupons by status

Closes #31562

## All Submissions:






## Background

The Woo mobile app is working on adding coupon management to the app. As part of that project, there are a few enhancements in the GET Coupon(s) API endpoint that we would need to make. 
- It looks like only published coupons are available in the API response. It would be nice to list all coupons available for the store, regardless of the status, since Core displays it that way. 

### Designs
<img width="1541" alt="Screenshot 2022-01-05 at 4 03 09 PM" src="https://user-images.githubusercontent.com/22608780/148203152-e532f716-8538-4ea2-a85a-e8267d6c329b.png">
<img width="292" alt="Screenshot 2022-01-05 at 4 03 52 PM" src="https://user-images.githubusercontent.com/22608780/148203270-9a1112a5-842f-4e3d-9cfb-152bc282da84.png">

## Proposed Changes
This PR adds the option to filter coupons by status (if requested in the API request). Using an example JSON, this is how it would look like:
- GET `/wc/v3/coupons` would return all coupons, regardless of the status. The default statuses I've seen are `publish`, `pending`, `draft`, `auto-draft`, `trash`.
- GET `wc/v3/coupons?status=publish, draft, pending` would return coupons with status as `publish`, `pending` or `draft`.

This change would allow us to reliably show the status of the coupon when displaying the list of coupons.

## How to test the changes in this Pull Request:
I have added a separate class `class-wc-rest-coupons-controller-tests.php` to test the coupon controller class. For manual testing: 

1. Add a few coupons in wp-admin → Coupons → Add New.
2. Publish some of the coupons, update some coupons to draft or pending.
3. In the API, run `GET /wp-json/wc/v3/coupons` and confirm that all the coupons (regardless of the status) is fetched in the API response.
4. In the API, run `GET /wp-json/wc/v3/coupons?status=publish, draft, pending` and confirm that only coupons with those statuses are fetched in the API response.

## Other information:




## Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

Adds the option to filter coupons by status when calling the `GET /coupons` endpoint. 

**This PR is in draft till [this PR](https://github.com/woocommerce/woocommerce/pull/31561) can be merged**
